### PR TITLE
Fixes for offset behaviour

### LINKED ISSUE #234: rocrand_set_offset unexpected behaviour
**Describe the bug**
`rorand_set_offset` seems to change the seed of the generator. The numbers produced by a generator with and without offset with the same seed are completely different.

**To Reproduce**
1. Install `rocrand-dev4.5.2` and `hip-dev4.5.2` version 4.5.2 using the package repositories
2. Compile the attached reproducer ([reprod.txt](https://github.com/ROCmSoftwarePlatform/rocRAND/files/7697563/reprod.txt)) with the following line:
`/opt/rocm-4.5.2/hip/bin/hipcc -I/opt/rocm-4.5.2/rocrand/include/ -L/opt/rocm-4.5.2/rocrand/lib/ -lrocrand test.cpp`

3. See that in the output no offset can be archived, but the two lists of random numbers seem to be completely independent:
```
0.0225561:0.750519
0.129137:0.628226
0.805372:0.446292
0.974561:0.55909
0.109374:0.951059
0.471769:0.231839
0.920535:0.930325
0.731697:0.620792
0.33033:0.228607
0.2921:0.888079
```

**Expected behavior**
Based on the description [here](https://github.com/ROCmSoftwarePlatform/rocRAND/blob/1afd81ce597b8fe5fcc0f976df061520e93a45dd/library/include/rocrand/rocrand.h#L515-L534), I would expect that the second list of random numbers generated is the same as the original one, but with a certain offset. For example in this case I would expect something like:
```
0.0225561:0.00121
0.129137:0.0225561
0.805372:0.129137
0.974561:0.805372
0.109374:0.974561
0.471769:0.109374
0.920535:0.471769
0.731697:0.920535
0.33033:0.731697
0.2921:0.33033
```
**Log-files**
[out.txt](https://github.com/ROCmSoftwarePlatform/rocRAND/files/7697581/out.txt)

**Environment**
[environment.txt](https://github.com/ROCmSoftwarePlatform/rocRAND/files/7697583/environment.txt)

**Additional context**
This issue was encountered while working on the rocrand backend for oneMKL.

### PR DESCRIPTION
**Pull request by Maetveis**
Fixes #234 for XORWOW, MRG32K3A, PHILOX4X32_10. Contains tests for SOBOL32 and SOBOL64 as well, no fix was required as these generators worked correctly.

Replaces the following PRs:
- #235
- #236
- #237
- #238

Additionally the gitlab-ci.yml ci script now pulls submodules (required because of the hiprand submodule).

### CODE REVIEW

**Comment by stanleytsang-amd** [comment]
@Maetveis I see there is one test failure on navi21 CI: test_rocrand_mrg32k3a_prng.  I am able to reproduce the error on my navi21 machine as well.  Do you see the same on your 6900xt?

[ RUN      ] rocrand_mrg32k3a_prng_offset/0.offsets_test
/home/AMD/statsang/rocRAND/test/test_rocrand_mrg32k3a_prng.cpp:550: Failure
Expected equality of these values:
  host_data0[i]
    Which is: 631382108
  host_data1[i + offset]
    Which is: 3430057079
[  FAILED  ] rocrand_mrg32k3a_prng_offset/0.offsets_test, where TypeParam = unsigned int (5 ms)
[----------] 1 test from rocrand_mrg32k3a_prng_offset/0 (5 ms total)

[----------] 1 test from rocrand_mrg32k3a_prng_offset/1, where TypeParam = float
[ RUN      ] rocrand_mrg32k3a_prng_offset/1.offsets_test
/home/AMD/statsang/rocRAND/test/test_rocrand_mrg32k3a_prng.cpp:550: Failure
Expected equality of these values:
  host_data0[i]
    Which is: 0.147005
  host_data1[i + offset]
    Which is: 0.798622
[  FAILED  ] rocrand_mrg32k3a_prng_offset/1.offsets_test, where TypeParam = float (5 ms)
[----------] 1 test from rocrand_mrg32k3a_prng_offset/1 (5 ms total)

[----------] Global test environment tear-down
[==========] 25 tests from 3 test suites ran. (5346 ms total)
[  PASSED  ] 23 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] rocrand_mrg32k3a_prng_offset/0.offsets_test, where TypeParam = unsigned int
[  FAILED  ] rocrand_mrg32k3a_prng_offset/1.offsets_test, where TypeParam = float

**Comment by Maetveis** [comment]
> @Maetveis I see there is one test failure on navi21 CI: test_rocrand_mrg32k3a_prng. I am able to reproduce the error on my navi21 machine as well. Do you see the same on your 6900xt?

Yes I can. The latest commit should fix it, by enabling assembly for navi.

File: library/include/rocrand/rocrand_common.h [hunk 1]
```diff
@@ -63,7 +63,8 @@ namespace detail {
       defined(__gfx904__) || \
       defined(__gfx906__) || \
       defined(__gfx908__) || \
-      defined(__gfx909__) )
+      defined(__gfx909__) || \
+      defined(__gfx1030__))
```
Comment (stanleytsang-amd) [inline library/include/rocrand/rocrand_common.h:67]: add gfx90a

**Review by stanleytsang-amd — CHANGES_REQUESTED** [review]

File: library/include/rocrand/rocrand_common.h (hunk 1 above)
Comment (Maetveis) [reply library/include/rocrand/rocrand_common.h:68]: Added.

**Comment by sbalint98** [comment]
Thank you, @Maetveis, for the PR; it seems to solve the issue described in issue #234

I encountered an issue when building the `develop_stream_20220110` branch from source. Whne using the cmake packages installed rocRAND, it seems to me that the `roc::rocrand` target does not advertise the necessary compile flags properly, and therefore fails to find the `rocrand.h` header. I find this issue very curious since this does not occur for the develop branch, and I cannot see any cmake related changes in this PR that would explain a difference.

I can reproduce the problem the following way, when on develop_stream_20220110

I am using the following minimal `test.cpp`:
```
#include<rocrand.h>
int main(){}
```
and CMakeLists.txt
```
cmake_minimum_required (VERSION 3.16)
project(test)
find_package(rocRAND REQUIRED)
add_executable(test test.cpp)
target_link_libraries(test roc::rocrand )
```

when configuring with:
`CXX=/opt/rocm-4.5.2/hip/bin/hipcc cmake -DCMAKE_PREFIX_PATH=/opt/rocm-4.5.2/lib/cmake/ -DrocRAND_DIR=<custom-install_path>/lib/cmake/rocrand/ ..`

I can reliably get an error when building due to the missing `rocrand.h` header.

Based on the compilation flags added by cmake:
`/opt/rocm-4.5.2/hip/bin/hipcc   -isystem /home/balint/rocRAND/build-develop-stream/install/include   -o CMakeFiles/test.dir/test.cpp.o -c /home/balint/rocrand-cmake-issuue-reprod/test.cpp` it looks like that the `isystem` path is incorrect since in case of my build the `rocrand.h` file is in the `include/rocrand` directory.

I configured rocRAND like this:
`CXX=/opt/rocm-4.5.2/hip/bin/hipcc  cmake -DBUILD_HIPRAND=OFF  -DCMAKE_INSTALL_PREFIX=$PWD/install  -DCMAKE_PREFIX_PATH=/opt/rocm-4.5.2/lib/cmake/  ..`

**Comment by stanleytsang-amd** [comment]
> Thank you, @Maetveis, for the PR; it seems to solve the issue described in issue #234
> 
> I encountered an issue when building the `develop_stream_20220110` branch from source. Whne using the cmake packages installed rocRAND, it seems to me that the `roc::rocrand` target does not advertise the necessary compile flags properly, and therefore fails to find the `rocrand.h` header. I find this issue very curious since this does not occur for the develop branch, and I cannot see any cmake related changes in this PR that would explain a difference.
> 
> I can reproduce the problem the following way, when on develop_stream_20220110
> 
> I am using the following minimal `test.cpp`:
> 
> ```
> #include<rocrand.h>
> int main(){}
> ```
> 
> and CMakeLists.txt
> 
> ```
> cmake_minimum_required (VERSION 3.16)
> project(test)
> find_package(rocRAND REQUIRED)
> add_executable(test test.cpp)
> target_link_libraries(test roc::rocrand )
> ```
> 
> when configuring with: `CXX=/opt/rocm-4.5.2/hip/bin/hipcc cmake -DCMAKE_PREFIX_PATH=/opt/rocm-4.5.2/lib/cmake/ -DrocRAND_DIR=<custom-install_path>/lib/cmake/rocrand/ ..`
> 
> I can reliably get an error when building due to the missing `rocrand.h` header.
> 
> Based on the compilation flags added by cmake: `/opt/rocm-4.5.2/hip/bin/hipcc -isystem /home/balint/rocRAND/build-develop-stream/install/include -o CMakeFiles/test.dir/test.cpp.o -c /home/balint/rocrand-cmake-issuue-reprod/test.cpp` it looks like that the `isystem` path is incorrect since in case of my build the `rocrand.h` file is in the `include/rocrand` directory.
> 
> I configured rocRAND like this: `CXX=/opt/rocm-4.5.2/hip/bin/hipcc cmake -DBUILD_HIPRAND=OFF -DCMAKE_INSTALL_PREFIX=$PWD/install -DCMAKE_PREFIX_PATH=/opt/rocm-4.5.2/lib/cmake/ ..`

Hello, I can chime in here.  The reason why you're seeing this problem with this PR is because this PR is likely based on a version of  rocRAND where we had introduced packaging changes - one of them being moving the rocrand headers from rocrand.h to rocrand/rocrand.h, as part of a upcoming change to ROCm's math libraries as a whole.  However this initiative ended up getting pushed back and thus in the develop branch we rolled back the packaging change, hence the develop branch works for you (and not this PR, where the changes have not been rolled back yet).  

I think once this PR is merged in you should be able to include rocRAND as normal - but let me verify that now for you.

**Comment by stanleytsang-amd** [comment]
@sbalint98 Confirmed that merging this PR with develop works with your example.  I think we can go ahead and merge this in.

**Review by stanleytsang-amd — APPROVED** [review]

**Pull request merged**